File: extension/chunker.js
/**
 * Sentence chunker — shared by the content script and the popup.
 *
 * Loaded as a classic script (not a module) so both runtimes can use the same
 * copy: content scripts share one isolated-world global, and popup.html pulls it
 * in with a plain <script> tag.
 *
 * Mirrors app/chunker.py. Returns spans into the source string rather than
 * copies, because the caller maps those offsets back onto DOM ranges to
 * highlight the sentence currently being spoken.
 */
(function (root) {
  const SENTENCE_ENDERS = new Set([".", "!", "?", "…"]);
  const TRAILING = new Set([
    '"', "'", "”", "’", ")", "]", "}", "…", "!", "?", ".",
  ]);
  const MAX_CHUNK = 280;

  const isSpace = (ch) => /\s/.test(ch);

  /** @returns {Array<[number, number]>} [start, end) spans covering the content. */
  function sentenceSpans(text, maxLen) {
    maxLen = maxLen || MAX_CHUNK;
    const spans = [];
    const n = text.length;
    let i = 0;

    while (i < n) {
      while (i < n && isSpace(text[i])) i++;
      if (i >= n) break;

      const start = i;
      let end = null;
      let lastSpace = -1;
      let j = i;

      while (j < n) {
        const ch = text[j];

        if (SENTENCE_ENDERS.has(ch)) {
          let k = j + 1;
          while (k < n && TRAILING.has(text[k])) k++;
          if (k >= n || isSpace(text[k])) {
            end = k;
            break;
          }
        }

        if (ch === "\n") {
          // A lone newline is a wrap; a blank line is a paragraph break.
          let k = j + 1;
          while (k < n && (text[k] === " " || text[k] === "\t" || text[k] === "\r")) k++;
          if (k < n && text[k] === "\n") {
            end = j;
            break;
          }
        }

        if (isSpace(ch)) lastSpace = j;

        if (j - start >= maxLen) {
          end = lastSpace > start ? lastSpace : j;
          break;
        }

        j++;
      }

      if (end === null) end = n;
      spans.push([start, end]);
      i = end;
    }

    return spans;
  }

  /**
   * @returns {Array<{start: number, end: number, text: string}>}
   * Spans whose trimmed text is non-empty, with the text carried along.
   */
  function chunks(text, maxLen) {
    const out = [];
    for (const [start, end] of sentenceSpans(text, maxLen)) {
      const piece = text.slice(start, end).trim();
      if (piece) out.push({ start, end, text: piece });
    }
    return out;
  }

  root.ReadAloudChunker = { sentenceSpans, chunks, MAX_CHUNK };
})(typeof self !== "undefined" ? self : this);

File: app/chunker.py
"""Split text into speakable chunks and report where each one sits in the source.

The GUI highlights the chunk currently being spoken, so every chunk is returned as
a (start, end) character span into the original string rather than as a copy.
"""

from __future__ import annotations

SENTENCE_ENDERS = ".!?…"
# Closing punctuation that belongs to the sentence it trails.
TRAILING = "\"'”’)]}…!?."
MAX_CHUNK = 280


def sentence_spans(text: str, max_len: int = MAX_CHUNK) -> list[tuple[int, int]]:
    """Return (start, end) spans covering the non-whitespace content of `text`.

    Breaks on sentence-ending punctuation, on blank lines, and — for runaway text
    with no punctuation — at the last word boundary before `max_len`.
    """
    spans: list[tuple[int, int]] = []
    i, n = 0, len(text)

    while i < n:
        while i < n and text[i].isspace():
            i += 1
        if i >= n:
            break

        start = i
        end = None
        last_space = -1
        j = i

        while j < n:
            ch = text[j]

            if ch in SENTENCE_ENDERS:
                k = j + 1
                while k < n and text[k] in TRAILING:
                    k += 1
                if k >= n or text[k].isspace():
                    end = k
                    break

            if ch == "\n":
                # Blank line = paragraph break. A single newline is just a wrap.
                k = j + 1
                while k < n and text[k] in " \t\r":
                    k += 1
                if k < n and text[k] == "\n":
                    end = j
                    break

            if ch.isspace():
                last_space = j

            if j - start >= max_len:
                end = last_space if last_space > start else j
                break

            j += 1

        if end is None:
            end = n
        spans.append((start, end))
        i = end

    return spans


def chunks(text: str, max_len: int = MAX_CHUNK) -> list[tuple[int, int, str]]:
    """Same as `sentence_spans` but carries the chunk text along."""
    out = []
    for start, end in sentence_spans(text, max_len):
        piece = text[start:end].strip()
        if piece:
            out.append((start, end, piece))
    return out
